src/backend/distributed/connection/connection_management.c

```diff
@@ -492,6 +492,12 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
 			continue;
 		}
 
+		if (RemoteConnectionClosed(connection))
```

Comment:
this check seems quite expensive

Reply:
Do you have any specific concerns? Doing some single shard pgbench tests, I do not see any regressions locally. Though, I should re-run this on HSC where PG15 is not supported yet.

We are at most adding 2 re-build per cached connection / per command. It might add some overhead, but if so I'm planning to move waitEventSet and events to MultiConnection so that we don't need to rebuild every-time. We can just call `WaitEventSetWait` on already allocated. Though, I'm not convinced that is needed yet.  I refrain doing that because then we'd have other edge cases like what if the socket of the connection has changed etc.

Btw, I remember we had some perf issues on re-building the WaitEventSet. However, I also remember that it was due to excessive rebuilding, see this discussion: https://github.com/citusdata/citus/pull/2783#pullrequestreview-253090173 After this patch, I don't think waitEvents are causing perf issues anymore. 

ps: I removed 1ms timeout, we don't need that.

Reply:
> We are at most adding 2 re-build per cached connection / per command.

In fact, 1 is enough, we don't need that in Shutdown, we can use `connection->forceCloseAtTransactionEnd ` flag

Reply:
For the performance, I did a test on Azure VMs (and correctness works as expected). I think the difference is in the order of noise. Do you agree the following is an OK test or do you have something different in mind?

- On an empty reference table such that the rest of the overhead is minimal
- pgbench -f /tmp/f.sqll -c 1 -j 1 -T 360 -P 1
- f.sqll: SELECT * FROM test;

main:
```
duration: 360 s
latency average = 0.873 ms
latency stddev = 0.616 ms
initial connection time = 6.192 ms
tps = 1145.482382 (without initial connection time)
```

This branch:
```
duration: 360 s
latency average = 0.852 ms
latency stddev = 0.680 ms
initial connection time = 3.151 ms
tps = 1173.761812 (without initial connection time)

```

Reply:
Was the reference table also on the coordinator?

Reply:
> Was the reference table also on the coordinator?

nope :)